pkg/services/ngalert/migration/models/models.go

```diff
@@ -0,0 +1,104 @@
+package models
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/util"
+)
+
+// MaxIncrementDeduplicationAttempts is the maximum number of attempts to try to deduplicate a string by appending
+// an incrementing suffix.
+const MaxIncrementDeduplicationAttempts = 10
+
+// Deduplicator is a utility for deduplicating strings. It keeps track of the strings it has seen and appends a unique
+// suffix to strings that have already been seen. It can optionally truncate strings before appending the suffix to
+// ensure that the resulting string is not longer than maxLen.
+// This implementation will first try to deduplicate via a sequential index suffix of the form " #2", " #3", etc.
+// If after MaxIncrementDeduplicationAttempts attempts it still cannot find a unique string, it will generate a new
+// unique uid and append that to the string.
+type Deduplicator struct {
+	set             map[string]int
+	caseInsensitive bool
+	maxLen          int
+	uidGenerator    func() string
+}
+
+// NewDeduplicator creates a new deduplicator.
+// caseInsensitive determines whether the string comparison should be case-insensitive.
+// maxLen determines the maximum length of deduplicated strings. If the deduplicated string would be longer than
+// maxLen, it will be truncated.
+func NewDeduplicator(caseInsensitive bool, maxLen int) *Deduplicator {
+	return &Deduplicator{
+		set:             make(map[string]int),
+		caseInsensitive: caseInsensitive,
+		maxLen:          maxLen,
+		uidGenerator:    util.GenerateShortUID,
+	}
+}
+
+// Deduplicate returns a unique string based on the given base string. If the base string has not already been seen by
+// this deduplicator, it will be returned as-is. If the base string has already been seen, a unique suffix will be
+// appended to the base string to make it unique.
+func (s *Deduplicator) Deduplicate(base string) (string, error) {
+	if s.maxLen > 0 && len(base) > s.maxLen {
+		base = base[:s.maxLen]
+	}
+	cnt, ok := s.contains(base)
+	if !ok {
+		s.add(base, 0)
+		return base, nil
+	}
+
+	// Start at 2, so we get a, a_2, a_3, etc.
+	for i := 2 + cnt; i < 2+cnt+MaxIncrementDeduplicationAttempts; i++ {
+		dedup := s.appendSuffix(base, fmt.Sprintf(" #%d", i))
+		if _, ok := s.contains(dedup); !ok {
+			s.add(dedup, 0)
+			return dedup, nil
+		}
+	}
+
+	// None of the simple suffixes worked, so we generate a new uid. We try a few times, just in case but this should
+	// almost always create a unique string on the first try.
+	for i := 0; i < 5; i++ {
```

Comment:
Perhaps we can reuse constant MaxIncrementDeduplicationAttempts here by slightly renaming it. Not a blocker at all. Just for the sake of not using magic literals in code :)